esx_service/utils/auth_data.py

```diff
@@ -417,14 +420,223 @@ def remove_datastore_access_privileges(self, conn, datastore_url):
+    - keeps a connetion to in-mem DB to prevent it from cleaning up static cache
+    - on first enable, check we are not enabled, populate the DB
+    - gets NEW connection when asked: conn = cache.new_connection()
+    - on refresh, drains in flight Auth managers using the hshared cache,
+       drop  connection (that will deleted the DB when sqlite refcount drops to 0) and recreate cache
+    """
+
+    # this is in-memory shared DB URL
+    MEMDB_NAME = "file::memory:?cache=shared"
+
+    # - How often to refresh memdb from on-disk DB.  Default 15 min.
+    REFRESH_SEC = int(os.getenv("VMDKOPS_MEMDB_REFRESH_SEC", default=900))
+
+
+    def __init__(self):
+        # Init connection to in-memory DB.
+        # We keep one connection open to assure the DB is not garbage collected too early
+        # We refresh it every MEMDB_REFRESH_SEC i. When we refresh, we first drain all in-flight
+        # AuthernticatoinDataManagers to make sure nobody is using the copy.
+        #
+        # 'None' means the feature is disabled
+        self._memdb_con = None
+
+        # Init Barrier
+        # We need to avoid refresh (which could be scheduled by Timer on a different thread)
+        # from messing DB content for auth. mgrs in flight. So  we acquire a barrier,
+        # and wait until all auth.managers in flight land.
+        self._barrier = threading.Condition()
+        self._in_flight = 0
+
+        # Init timer for periodic refreshes
+        self._timer = None
+
+    def __del__(self):
+        pass
+
+    def _drain_and_close(self):
+        """
+        Drains Auth. managers in fight and closes last memdb connection.
+        *** EXPECTS to be run under _barrier lock.
+        """
+        logging.debug("_Drain and close")
+        # drain the in-flight users
+        self._memdb_con.close()
+        self._memdb_con = None
+        while self._in_flight > 0:
+            logging.debug("Wating for inflight %d", self._in_flight)
+            self._barrier.wait()
+        logging.debug("DONE")
+
+
+    def enable(self):
+        """
+        Enables in-memory ReadOnly cache DB for auth DB.
+        This includes opening on-disk DB and populating in-memory DB from it.
+        Note this will force the connection to treat the DB as read only.
+        """
+        logging.debug("memdb_enable")
+        if not self.is_enabled():
+            logging.debug("memdb_enable Starting refresh")
+            # No need to drain anything as nobody was using the cache so far
+            # Proceed to creation os a new cache
+            self._memdb_refresh_periodically(frequency_sec=self.__class__.REFRESH_SEC)
+
+    def disable(self):
+        """
+        Disables the DB cache and closes the connection. The  DB will be deleted when all
+        AuthorizationDataManagers using this cache land and close their connection.
+        """
+        logging.debug("memdb_disable")
+        with self._barrier:
+            if self.is_enabled():
```

Comment:
This call tries to re-acquire the lock: `with self._barrier`. I'm not sure if these locks are re-entrant. If not, this is a deadlock.